test/integration/buffer_accounting_integration_test.cc

```diff
@@ -0,0 +1,281 @@
+  const uint32_t upstream_h2_stream_window = 64 * 1024;
+  const uint32_t upstream_h2_conn_window = 1024 * 1024 * 1024; // Effectively unlimited
+  const uint32_t request_body_size = 4096;
+  const uint32_t response_body_size = 4096;
+
+  initializeWithBufferConfig({connection_watermark, downstream_h2_stream_window,
+                              downstream_h2_conn_window, upstream_h2_stream_window,
+                              upstream_h2_conn_window},
+                             num_requests);
+  buffer_factory_->setExpectedAccountBalance(request_body_size, num_requests);
+
+  // Makes us have Envoy's writes to upstream return EAGAIN
+  writev_matcher_->setDestinationPort(fake_upstreams_[0]->localAddress()->ip()->port());
+  writev_matcher_->setWritevReturnsEgain();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  auto responses = sendRequests(num_requests, request_body_size, response_body_size);
+
+  // Wait for all requests to have accounted for the requests we've sent.
+  if (streamBufferAccounting()) {
+    EXPECT_TRUE(
+        buffer_factory_->waitForExpectedAccountBalanceWithTimeout(TestUtility::DefaultTimeout))
+        << "buffer total: " << buffer_factory_->totalBufferSize()
+        << " buffer max: " << buffer_factory_->maxBufferSize() << printAccounts();
+  }
+
+  writev_matcher_->setResumeWrites();
+
+  for (auto& response : responses) {
+    ASSERT_TRUE(response->waitForEndStream());
+    ASSERT_TRUE(response->complete());
+  }
+}
+
+TEST_P(HttpBufferWatermarksTest, ShouldTrackAllocatedBytesToDownstream) {
+  const int num_requests = 5;
+  const uint32_t connection_watermark = 32768;
+  const uint32_t downstream_h2_stream_window = 512 * 1024;
+  const uint32_t downstream_h2_conn_window = 64 * 1024;
+  const uint32_t upstream_h2_stream_window = 64 * 1024;
+  const uint32_t upstream_h2_conn_window = 1024 * 1024 * 1024; // Effectively unlimited
+  const uint32_t request_body_size = 4096;
+  const uint32_t response_body_size = 16384;
+
+  initializeWithBufferConfig({connection_watermark, downstream_h2_stream_window,
+                              downstream_h2_conn_window, upstream_h2_stream_window,
+                              upstream_h2_conn_window},
+                             num_requests);
+  buffer_factory_->setExpectedAccountBalance(response_body_size, num_requests);
+  writev_matcher_->setSourcePort(lookupPort("http"));
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Simulate TCP push back on the Envoy's downstream network socket, so that outbound frames
+  // start to accumulate in the transport socket buffer.
+  writev_matcher_->setWritevReturnsEgain();
+
+  auto responses = sendRequests(num_requests, request_body_size, response_body_size);
+
+  // Wait for all requests to buffered the response from upstream.
+  if (streamBufferAccounting()) {
+    EXPECT_TRUE(
+        buffer_factory_->waitForExpectedAccountBalanceWithTimeout(TestUtility::DefaultTimeout))
```

Comment:
Not sure that sending requests should really be the job of the buffer_factory. That seems a bit strange.

Reply:
This comment is about the wait with timeout API provided by the buffer factory.

Reply:
Perhaps we're miscommunicating.

I see the issue as follows: 
1. The test thread has to send the requests
2. The test thread also has to set these expectations and wait on them

If the test thread runs (1) followed by (2), there's a possibility of a race causing the wait to timeout even though accounts are correctly sized due to the accounts notifying if expectations are met in the worker when updating accounts. i.e. if all accounts are updated, and we then set the expectation it won't fire since none of them have any additional events.

So instead this is split into
1. Send request
2. Set expectation
3. Wait for expectation

And we can do 2, 1, and 3. Alternatively we can push those into a single function, as what I think you'd suggest would require, but it seems strange for the buffer factory to be involved with 2.

Reply:
Duplicating the internal information of the account doesn't seem right -- seems like a leaky abstraction. 

I also prefer using the notification as IIRC `AwaitWithTimeout` doesn't have a signaling mechanism while the notification does, awaking blocked threads.

Reply:
AwaitWithTimeout checks every time that the mutex is released.  But there's a requirement that updates happen when the mutex is held.

Reply:
Anyway, I think we can start with this test framework API and improve it in the future as needed.